lib/dao.js

```diff
@@ -586,6 +586,195 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+      cb = options;
+      options = {};
+    }
+  }
+
+  cb = cb || utils.createPromiseCallback();
+  data = data || {};
+  options = options || {};
+
+  assert(typeof data === 'object', 'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+
+  var hookState = {};
+
+  var self = this;
+  var Model = this;
+  var connector = Model.getConnector();
+
+  var id = getIdValue(this, data);
+  if (id === undefined || id === null) {
+    return this.create(data, options, cb);
+  }
+
+  var context = {
+    Model: Model,
+    query: byIdQuery(Model, id),
+    hookState: hookState,
+    options: options
+  };
+  Model.notifyObserversOf('access', context, doReplaceOrCreate);
+
+  function doReplaceOrCreate(err, ctx) {
+    if (err) return cb(err);
+
+    var isOriginalQuery = isWhereByGivenId(Model, ctx.query.where, id);
+    if (connector.replaceOrCreate && isOriginalQuery) {
+      var context = {
+        Model: Model,
+        where: ctx.query.where,
+        data: data,
+        hookState: hookState,
+        options: options
+      };
+      Model.notifyObserversOf('before save', context, function(err, ctx) {
+        if (err) return cb(err);
+
+        data = ctx.data;
+        var update = data;
+        var inst = data;
+        if (!(data instanceof Model)) {
+          inst = new Model(data, { applyDefaultValues: false });
+        }
+        update = inst.toObject(false);
+
+        Model.applyProperties(update, inst);
+        Model = Model.lookupModel(update);
+
+        var connector = self.getConnector();
+
+        // TODO: validateReplaceOrCreate needs to be set to true by default in model JSON
+        if (Model.settings.validateReplaceOrCreate === false) {
+          callConnector();
+        } else {
+          inst.isValid(function(valid) {
+            if (!valid) {
+              if (Model.settings.validateReplaceOrCreate) {
+                return cb(new ValidationError(inst), inst);
+              } else {
+                // TODO(bajtos) Remove validateReplaceOrCreate:undefined in v3.0 ?!
+                console.warn('Ignoring validation errors in replaceOrCreate():');
+                console.warn('  %s', new ValidationError(inst).message);
+                // continue with replaceOrCreate
+              }
+            }
+            callConnector();
+          }, update);
+        }
+
+        function callConnector() {
+          update = removeUndefined(update);
+          context = {
+            Model: Model,
+            where: ctx.where,
+            data: update,
+            currentInstance: inst,
+            hookState: ctx.hookState,
+            options: options
+          };
+          Model.notifyObserversOf('persist', context, function(err) {
+            if (err) return done(err);
+            if (connector.replaceOrCreate.length === 4) {
```

Comment:
Since this is a new connector operation and we don't have to preserve backwards compatibility, you can always assume the connector method takes 4 arguments and get rid of this `if` check.